tests/ops/op_math/test_composite.py

```diff
@@ -120,12 +120,30 @@ def test_ndim_params_raises_error(self):
         with pytest.raises(AttributeError):
             _ = op.ndim_params
 
-    def test_batch_size_raises_error(self):
-        """Test that calling batch_size raises a ValueError."""
-        op = ValidOp(*self.simple_operands)
-
-        with pytest.raises(AttributeError):
-            _ = op.batch_size
+    def test_batch_size_None(self):
+        """Test that the batch size is none if no operands have batching."""
+        prod_op = ValidOp(qml.PauliX(0), qml.RX(1.0, wires=0))
+        assert prod_op.batch_size is None
+
+    def test_batch_size_all_batched(self):
+        """Test that the batch_size is correct when all operands are batched."""
+        base = qml.RX(np.array([1.2, 2.3, 3.4]), 0)
+        op = ValidOp(base, base, base)
+        assert op.batch_size == 3
+
+    def test_batch_size_not_all_batched(self):
+        """Test that the batch_size is correct when all operands are not batched."""
+        base = qml.RX(np.array([1.2, 2.3, 3.4]), 0)
+        op = ValidOp(base, qml.RY(1, 0), qml.RZ(np.array([1, 2, 3]), wires=2))
```

Comment:
the `RY` would ultimately be converted into a batch_size-3 equivalent if this was a real composite operation, right? If so, whose responsibility is it to ensure it, and should the CompositeOp implementation know whose responsibility it is/if it's upheld later? eg. prod takes care of the matrix work in this PR, but what if it didn't?

Reply:
To avoid changing classes dynamically, `RY` is not converted into a batch_size-3 equivalent:
```pycon
>>> base = qml.RX(np.array([1.2, 2.3, 3.4]), 0)
>>> qml.prod(base, qml.RY(1, 0), qml.RZ(np.array([1, 2, 3]), wires=2))
RX(array([1.2, 2.3, 3.4]), wires=[0]) @ RY(1, wires=[0]) @ RZ(array([1, 2, 3]), wires=[2])
```

Right now, the `matrix` method is responsible to ensure batching is available. If I'm correct, the whole batching pipeline doesn't have a way to determine if batching is available for an operator or not. Thus if you use batching with an operator that doesn't support it, either you get a weird error, or you get the wrong result.

Reply:
But again, I'm sure @dwierichs has a better answer

Reply:
>  If so, whose responsibility is it to ensure it,

An operation that claims to support batching should, well, support batching. In particular that means that the `matrix` method of the operation needs to know what to do in case broadcasting/batching is used.
For `CompositeOp`s, this in particular means that it has to support a variety of combinations (all input ops batched, some batched, none batched...).

When executing a quantum tape, for example, no conversion is necessary, because devices that support broadcasting will use that matrix multiplication can be broadcasted itself.

Note that there is `qml.ops.qubit.attributes.supports_broadcasting` which contains all operations that support broadcasting. I think this PR should add the updated operations to that attribute :)

Reply:
> Thus if you use batching with an operator that doesn't support it, either you get a weird error, or you get the wrong result.

That's right. Typically you will get an error because of dimension mismatches. This is not special for broadcasting though: If we want to ensure operations are not instantiated with ill-formatted inputs, the operations have to check themselves that the input is okay.

Reply:
Added `Prod` and `Sum` into `qml.ops.qubit.attributes.supports_broadcasting`